pkg/authentication/scram/scram.go

```diff
@@ -47,9 +81,238 @@ func EnsureAgentSecret(getUpdateCreator secret.GetUpdateCreator, secretNsName ty
 		agentSecret.Data[AgentKeyfileKey] = []byte(generatedContents)
 	}
 
-	return automationConfigModification(
-		string(agentSecret.Data[AgentPasswordKey]),
-		string(agentSecret.Data[AgentKeyfileKey]),
-		[]automationconfig.MongoDBUser{},
-	), getUpdateCreator.UpdateSecret(agentSecret)
+	return automationConfigModification(string(agentSecret.Data[AgentPasswordKey]), string(agentSecret.Data[AgentKeyfileKey]), desiredUsers), secret.CreateOrUpdate(secretGetUpdateCreateDeleter, agentSecret)
+}
+
+// ensureScramCredentials will ensure that the ScramSha1 & ScramSha256 credentials exist and are stored in the credentials
+// secret corresponding to user of the given MongoDB deployment.
+func ensureScramCredentials(getUpdateCreator secret.GetUpdateCreator, user mdbv1.MongoDBUser, mdb mdbv1.MongoDB) (scramcredentials.ScramCreds, scramcredentials.ScramCreds, error) {
+	passwordKey := user.PasswordSecretRef.Key
+	if passwordKey == "" {
+		passwordKey = defaultPasswordKey
+	}
+	password, err := secret.ReadKey(getUpdateCreator, passwordKey, types.NamespacedName{Name: user.PasswordSecretRef.Name, Namespace: mdb.Namespace})
+	if err != nil {
+		// if the password is deleted, that's fine we can read from the stored credentials that were previously generated
+		if errors.IsNotFound(err) {
+			zap.S().Debugf("password secret was not found, reading from credentials from secret/%s", scramCredentialsSecretName(mdb.Name, user.Name))
```

Comment:
Yes it can definitely be cleaned up. I really don't like the idea of adding the logger to the method signature as it would make it less importable from other packages. 

However maybe this is the best thing to do as it allows us to easily distinguish between different resources using a `SugaredLogger`. I would say if we do vote for this, we should ensure the logging is consistent all throughout the application.